### DISCUSSION TITLE
Running Avalonia in Docker image

### DISCUSSION THREAD

**Discussion by elveejay1** [AvaloniaUI/Avalonia#8915] [Q&A]
I tried to create a docker image, but ran into a problem:
I get an exception:
'libX11.so.6' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibX11.so.6: cannot open shared object file: No such file or directory'

When I try to run my code on a "real" linux VM (in Hyper V) things work fine.

My Docker file looks like this:

```
#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.

FROM mcr.microsoft.com/dotnet/runtime:6.0 AS base
WORKDIR /app

FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /src
COPY ["SampleApplication/SampleApplication.csproj", "SampleApplication/"]
RUN dotnet restore "SampleApplication/SampleApplication.csproj"
COPY . .
WORKDIR "/src/SampleApplication"
RUN dotnet build "SampleApplication.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "SampleApplication.csproj" -c Release -o /app/publish --runtime linux-x86 --self-contained=true

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "SampleApplication.dll"]

```

Has anyone tried this? I'm guessing that I might need to run some installations first. I know others have had similar issues (https://github.com/AvaloniaUI/Avalonia/issues/3020#issuecomment-533881126) though I'm not sure if that solution applies in my case, or if so, how to implement it (I know next to nothing about linux).

**Comment by kekekeks** [comment]
You probably need to follow [this guide](https://dev.to/darksmile92/run-gui-app-in-linux-docker-container-on-windows-host-4kde) (first link from "running x11 apps inside docker windows" google search)

**Reply by elveejay1** [reply]
Thanks for the quick response :-) 

Hm....not too happy about the line
set-variable -name DISPLAY -value YOUR-IP:0.0
as we're multiple devs, and we all have dynamic IP.

I'd hoped that I could run some apt-get install <whatever> lines and that was it.

I hope to have everything in just the docker file, so it will be easier for other devs to run the program.

**Reply by elveejay1** [reply]
I tried adding
RUN apt-get update -y
RUN apt-get install -y libx11-dev
RUN apt-get install -y libx11-6:amd64
before my dotnet restore. Still get the same error, though, but maybe there's something else that I'm missing?

**Reply by SuperJMN** [reply]
I'm interested in this, too.
I'd like to run my application inside a Kubernetes cluster and saw there are problems with it.
Can you please tell me if there's new information on the topic?

**Comment by memtech3** [comment]
Try running `sudo apt-get update && apt-get install -y x11-apps` command, then run `dotnet restore`. It worked for me on Docker on WSL2 on Windows 11

Picked out of https://askubuntu.com/questions/1249043/run-simple-x11-app-in-docker-container-on-ubuntu-20-04

**Comment by lindexi** [comment]
I use the Xvfb in docker:

```
FROM dotnet-runtime:10.0.5


RUN apt-get install -y xvfb
ENV DISPLAY=:0


RUN apt install -y fonts-ipafont
RUN apt install -y fonts-noto fonts-ubuntu fonts-roboto
RUN fc-cache -fv


RUN apt-get clean

````

Run the code before `BuildAvaloniaApp`:

```csharp
            var display = Environment.GetEnvironmentVariable("DISPLAY");
            Process.Start("Xvfb",
            [
                display,
                "-screen", "0",
                "1920x1080x24"
            ]);

            //Console.WriteLine($"Start xvfb");

            var stopwatch = Stopwatch.StartNew();
            var spinWait = new SpinWait();
            while (stopwatch.Elapsed < TimeSpan.FromSeconds(5))
            {
                var displayHandler = XOpenDisplay(IntPtr.Zero);
                //Console.WriteLine($"XOpenDisplay {displayHandler}");
                if (displayHandler == 0)
                {
                    spinWait.SpinOnce(10);
                }
                else
                {
                    Console.WriteLine($"Wait {stopwatch.Elapsed}");
                    XCloseDisplay(displayHandler);
                    return;
                }
            }

            throw new NotSupportedException($"Can not start X11");
```
